### PR TITLE
[python-package] fix mypy errors in engine.py

### PR DESCRIPTION
**Pull request by jameslamb**
Contributes to #3867.

This PR fixes `mypy` errors in `engine.py`

> python-package/lightgbm/engine.py:226: error: Incompatible types in assignment (expression has type "Set[<nothing>]", variable has type "Optional[List[Callable[..., Any]]]")
python-package/lightgbm/engine.py:230: error: Incompatible types in assignment (expression has type "Set[Callable[..., Any]]", variable has type "Optional[List[Callable[..., Any]]]")
python-package/lightgbm/engine.py:242: error: Item "List[Callable[..., Any]]" of "Optional[List[Callable[..., Any]]]" has no attribute "add"
python-package/lightgbm/engine.py:242: error: Item "None" of "Optional[List[Callable[..., Any]]]" has no attribute "add"
python-package/lightgbm/engine.py:244: error: Item "List[Callable[..., Any]]" of "Optional[List[Callable[..., Any]]]" has no attribute "add"
python-package/lightgbm/engine.py:244: error: Item "None" of "Optional[List[Callable[..., Any]]]" has no attribute "add"
python-package/lightgbm/engine.py:247: error: Item "List[Callable[..., Any]]" of "Optional[List[Callable[..., Any]]]" has no attribute "add"
python-package/lightgbm/engine.py:247: error: Item "None" of "Optional[List[Callable[..., Any]]]" has no attribute "add"
python-package/lightgbm/engine.py:252: error: Item "List[Callable[..., Any]]" of "Optional[List[Callable[..., Any]]]" has no attribute "add"
python-package/lightgbm/engine.py:252: error: Item "None" of "Optional[List[Callable[..., Any]]]" has no attribute "add"
python-package/lightgbm/engine.py:254: error: Item "None" of "Optional[List[Callable[..., Any]]]" has no attribute "__iter__" (not iterable)
python-package/lightgbm/engine.py:255: error: Unsupported left operand type for - ("List[Callable[..., Any]]")
python-package/lightgbm/engine.py:255: error: Unsupported left operand type for - ("None")
python-package/lightgbm/engine.py:255: note: Left operand is of type "Optional[List[Callable[..., Any]]]"
python-package/lightgbm/engine.py:256: error: Incompatible types in assignment (expression has type "List[Union[Callable[..., Any], Any]]", variable has type "Set[Union[Callable[..., Any], Any]]")
python-package/lightgbm/engine.py:300: error: Incompatible types in assignment (expression has type "float", variable has type "List[Any]")

Confirmed that this suppresses these warnings by running the following.

```shell
mypy \
    --exclude='python-package/compile/|python-package/build' \
    --ignore-missing-imports \
    python-package/
```

### Notes for Reviewers

Some of these errors caught an issue in the type hints for `callbacks.EarlyStopException`. That exception has a property `best_score` which seems like it should be a float, but in every place where one of those is created, `best_score` is populated with a list of eval results similar to this:

```python
[
    ('training', 'binary_logloss', 0.014361246252423975, False),
    ('training', 'mape', 0.013889796906138775, False),
    ('valid', 'binary_logloss', 0.0561150102648419, False),
    ('valid', 'mape', 0.042410104106485866, False)
]
```

https://github.com/microsoft/LightGBM/blob/67b4205c8043326553e294fa2c01ad1189784631/python-package/lightgbm/engine.py#L300

https://github.com/microsoft/LightGBM/blob/67b4205c8043326553e294fa2c01ad1189784631/python-package/lightgbm/callback.py#L280

https://github.com/microsoft/LightGBM/blob/67b4205c8043326553e294fa2c01ad1189784631/python-package/lightgbm/callback.py#L307

### CODE REVIEW

File: python-package/lightgbm/callback.py [hunk 1]
```diff
@@ -18,15 +18,15 @@ def _lt_delta(curr_score: float, best_score: float, delta: float) -> bool:
...
+    def __init__(self, best_iteration: int, best_score: List[Tuple[str, str, float, bool]]) -> None:
         """Create early stopping exception.
 
         Parameters
         ----------
         best_iteration : int
             The best iteration stopped.
-        best_score : float
-            The score of the best iteration.
+        best_score : list of (eval_name, metric_name, eval_result, is_higher_better) tuples
```
Comment (StrikerRUS) [inline python-package/lightgbm/callback.py:28]: Refer to #4575.
```suggestion
        best_score : list of (eval_name, metric_name, eval_result, is_higher_better) tuple
```

File: python-package/lightgbm/callback.py [hunk 2]
```diff
@@ -18,15 +18,15 @@ def _lt_delta(curr_score: float, best_score: float, delta: float) -> bool:
 class EarlyStopException(Exception):
     """Exception of early stopping."""
 
-    def __init__(self, best_iteration: int, best_score: float) -> None:
+    def __init__(self, best_iteration: int, best_score: List[Tuple[str, str, float, bool]]) -> None:
```
Comment (StrikerRUS) [inline python-package/lightgbm/callback.py:21]: It seems that this list can contain 5-element tuples.
https://github.com/microsoft/LightGBM/blob/67b4205c8043326553e294fa2c01ad1189784631/python-package/lightgbm/callback.py#L280
https://github.com/microsoft/LightGBM/blob/67b4205c8043326553e294fa2c01ad1189784631/python-package/lightgbm/callback.py#L275
https://github.com/microsoft/LightGBM/blob/67b4205c8043326553e294fa2c01ad1189784631/python-package/lightgbm/callback.py#L51

**Review by StrikerRUS — CHANGES_REQUESTED** [review]

File: python-package/lightgbm/callback.py (hunk 2 above)
Comment (jameslamb) [reply python-package/lightgbm/callback.py:21]: oooo thank you! I checked the format of this with a debugger, but forgot you could ask to show standard deviation.

I'll update this hint.

**Review by StrikerRUS — APPROVED** [review]
Thank you very much for fixing callback typehints!

**Pull request merged**